src/goto-programs/goto_coverage.cpp

```diff
@@ -38,64 +62,666 @@ void goto_coveraget::add_false_asserts(goto_functionst &goto_functions)
+  // noted that this should be done before the updating of *top_ptr
+  if (pre_cond_flg)
+  {
+    // change 'not (a && a) and b' back to 'a && a or b'
+    *root_ptr = old_root;
+  }
+
+  // 8. update *top_ptr;
+  join_expr.clear();
+  *top_ptr = old_top;
+  join_expr = exprt(op_tp, bool_type());
+  join_expr.operands().emplace_back(*top_ptr);
+  join_expr.operands().emplace_back(rhs);
+  *top_ptr = join_expr;
+
+  // 9. update rhs_ptr
+  rhs_ptr = top_ptr->operands().begin();
+  rhs_ptr++;
+}
+
+/*
+  - flag: if we have handled one
+*/
+void goto_coveraget::collect_operands(
+  const exprt &expr,
+  std::list<exprt> &operands,
+  bool &flag)
+{
+  const std::string &id = expr.id().as_string();
+
+  if ((id == "and" || id == "or") && flag == false)
+  {
+    bool flg0 = false;
+    collect_operands(expr.op0(), operands, flg0);
+    if (!flg0)
+      operands.emplace_back(expr.op0());
+    // operators.emplace_back(id);
+    bool flg1 = false;
+    collect_operands(expr.op1(), operands, flg1);
+    if (!flg1)
+      operands.emplace_back(expr.op1());
+    flag = true;
+  }
+  else
+  {
+    forall_operands (it, expr)
+      collect_operands(*it, operands, flag);
+  }
+}
+
+void goto_coveraget::collect_atom_operands(
+  const exprt &expr,
+  std::set<exprt> &atoms)
+{
+  const auto &id = expr.id();
+  forall_operands (it, expr)
+    collect_atom_operands(*it, atoms);
+  if (
+    id == exprt::equality || id == exprt::notequal || id == exprt::i_lt ||
+    id == exprt::i_gt || id == exprt::i_le || id == exprt::i_ge)
+  {
+    atoms.insert(expr);
+  }
+}
+
+void goto_coveraget::collect_operators(
+  const exprt &expr,
+  std::list<std::string> &operators)
+{
+  std::string str = from_expr(expr);
+  // prepocess: remove parenthesis in type_cast
+  // e.g. if(return_bool()) ==> IF !(_Bool)return_value$_return_bool$1
+  //                        ==> IF !return_value$_return_bool$1
+  // !FIXME: maybe this should be regex like \(\_[A-Za-z]+\) iF there are other typecasts
+  const std::string subStr = "(_Bool)";
+  size_t pos = str.find(subStr);
+  while (pos != std::string::npos)
+  {
+    str.erase(pos, subStr.length());
+    pos = str.find(subStr, pos);
+  }
+
+  std::list<std::string> opt;
+  for (std::size_t i = 0; i < str.length(); i++)
+  {
+    if (str[i] == ' ')
+      continue;
+    else if (str[i] == '|' && i + 1 < str.length() && str[i + 1] == '|')
+    {
+      opt.emplace_back("||");
+      i++;
+    }
+    else if (str[i] == '&' && i + 1 < str.length() && str[i + 1] == '&')
+    {
+      opt.emplace_back("&&");
+      i++;
+    }
+    else if (str[i] == '(')
+      opt.emplace_back("(");
+    else if (str[i] == ')')
+      opt.emplace_back(")");
+  }
+
+  // add implied parentheses in boolean expression
+  // e.g.
+  //   if(a&&b || c&&d) ==> if((a&&b) || (c&&d))
+  //   (a == 1 || b != 2 && (b == 3 || a == 1)) => (a == 1 || (b != 2 && (b == 3 || a == 1)))
+  // general rule: add parenthesis between || and &&
+  // (&&||&&)
```

Comment:
1. For what purpose do you add "implied parentheses"?
2. Does it work on constant_string expressions as well, e.g. `"a &&"[0] == 'a'`?
3. In general, again, hard-coded C assumption. Symbols could contain `(` just fine.

Reply:
1. Implied parentheses in a boolean expression refer to the implicit order of operations dictated by the precedence and associativity rules of boolean operators. E.g. `&&` > `||`.
2. Not sure. I guess it could. Need to do some experiment